include/circt/Dialect/FIRRTL/FIRRTLAnnotations.h

```diff
@@ -63,6 +76,40 @@ class AnnotationSet {
   DictionaryAttr
   getArgumentAttrDict(ArrayRef<NamedAttribute> otherPortAttrs = {}) const;
 
+  /// Store this annotation set on an operation. If `removeOnEmpty` is set, the
+  /// annotation attribute is removed from the operation altogether. Returns
+  /// true if the operation was modified, false otherwise.
+  bool updateOperation(Operation *op, bool removeOnEmpty = true) const;
+
+  /// Store this annotation set in the attribute list of an operation. If
+  /// `removeOnEmpty` is set, the annotation attribute is removed from the
+  /// operation altogether if this set is empty. Returns true if the attribute
+  /// list was modified, false otherwise.
+  bool updateAttrs(NamedAttrList &attrs, bool removeOnEmpty = true) const;
```

Comment:
This comment is out of date.  Please also describe what `removeOnEmpty` does here.

More importantly, it is also not obvious to me that this is setting the "annotations" attribute in `attrs` with the attribute set.  Could we make this be more port specific somehow?  The version immediately below looks more directed and obvious how to use it.  Do we need this one?

Reply:
The argument to have this as well was that if you are in the process of doing multiple modifications of an operation's attributes through a `NamedAttrList` (or you are building one), this would allow you to operate on that list directly, without having to first intern it as a `DictionaryAttr` just to modify it again through `updateDictionaryAttr` again immediately after.

I would love to see some attribute mutation classes in MLIR in general. Right now you can choose between `NamedAttrList`, rolling your own vector of `NamedAttribute`s and friends, with various degrees of not-modified-optimizations. Since the attributes are essentially an immutable data structure, a lot of the mutation API idioms we have for those could apply. Like a `MutableAttribute` helper to allow attributes to be mutated multiple types (and maybe in a nested fashion) before interning them into the context again :thinking:.